hera_mc/weather.py

```diff
@@ -0,0 +1,233 @@
+            raise ValueError('time must be an astropy Time object')
+        if variable not in weather_sensor_dict.keys():
+            raise ValueError('variable must be a key in weather_sensor_dict.')
+        if not isinstance(value, float):
+            raise ValueError('value must be a float.')
+        weather_time = floor(time.gps)
+
+        return cls(time=weather_time, variable=variable, value=value)
+
+
+@tornado.gen.coroutine
+def _helper_create_from_sensors(starttime, stoptime, variables=None):
+    """
+    Create a list of weather objects from sensor data using tornado server.
+
+    Parameters:
+    ------------
+    starttime: astropy time object
+        time to start getting history.
+    stoptime: astropy time object
+        time to stop getting history.
+    variable: string
+        variable to get history for. Must be a key in weather_sensor_dict,
+        defaults to all keys in weather_sensor_dict
+
+    Returns:
+    -----------
+    A list of WeatherData objects (only accessible via a yield call)
+    """
+    from katportalclient import KATPortalClient
+
+    if not isinstance(starttime, Time):
+        raise ValueError('starttime must be an astropy Time object')
+
+    if not isinstance(stoptime, Time):
+        raise ValueError('stoptime must be an astropy Time object')
+
+    if variables is None:
+        variables = weather_sensor_dict.keys()
+    elif not isinstance(variables, (list, tuple)):
+        variables = [variables]
+
+    sensor_names = []
+    sensor_var_dict = {}
+    for var in variables:
+        if var not in weather_sensor_dict.keys():
+            raise ValueError('variable must be a key in weather_sensor_dict')
+        sensor_names.append(weather_sensor_dict[var]['sensor_name'])
+        sensor_var_dict[weather_sensor_dict[var]['sensor_name']] = var
+
+    portal_client = KATPortalClient(katportal_url, on_update_callback=None)
+
+    histories = yield portal_client.sensors_histories(sensor_names, starttime.unix,
+                                                      stoptime.unix, timeout_sec=60)
+
+    weather_obj_list = []
+    for sensor_name, history in histories.items():
+        times = []
+        values = []
+        variable = sensor_var_dict[sensor_name]
+        for item in history:
+            # status is usually nominal, but can indicate sensor errors.
+            # Since we can't do anything about those and the data might be bad, ignore them
+            if item.status != 'nominal':
+                continue
+
+            # the value_timestamp is the sensor timestamp, while the other is
+            # when the recording system got it. The value_timestamp isn't always
+            # present, so test for it
+            if 'value_timestamp' in item._fields:
+                timestamp = item.value_timestamp
+            else:
+                timestamp = item.timestamp
+            # sometimes there are duplicates. Protect against that
+            if timestamp not in times:
```

Comment:
If this is what we're doing, `times` should be a `dict` or a `set`, for which the `x in y` test will be much more efficient. Seeing as the sensor data comes in at a high cadence this is one of the few times when the performance implications of that kind of test might have non-negligible implications.

Reply:
ok, I turned the times and values into a dict.